fluss-flink/fluss-flink-common/src/main/java/com/alibaba/fluss/flink/tiering/source/enumerator/TieringSourceEnumerator.java

```diff
@@ -0,0 +1,470 @@
+                new ClientMetricGroup(metricRegistry, "LakeTieringService");
+        this.rpcClient = RpcClient.create(flussConf, clientMetricGroup);
+        MetadataUpdater metadataUpdater = new MetadataUpdater(flussConf, rpcClient);
+        this.coordinatorGateway =
+                GatewayClientProxy.createGatewayProxy(
+                        metadataUpdater::getCoordinatorServer, rpcClient, CoordinatorGateway.class);
+        this.splitGenerator = new TieringSplitGenerator(flussAdmin);
+
+        LOG.info("Starting register Tiering Service to Fluss Coordinator...");
+        try {
+            LakeTieringHeartbeatResponse heartbeatResponse =
+                    waitHeartbeatResponse(
+                            coordinatorGateway.lakeTieringHeartbeat(basicHeartBeat()));
+            this.flussCoordinatorEpoch = heartbeatResponse.getCoordinatorEpoch();
+            LOG.info(
+                    "Register Tiering Service to Fluss Coordinator(epoch={}) success.",
+                    flussCoordinatorEpoch);
+
+        } catch (Exception e) {
+            LOG.error("Register Tiering Service failed due to ", e);
+            throw new FlinkRuntimeException("Register Tiering Service failed due to ", e);
+        }
+        this.context.callAsync(
+                this::requestTieringTableSplitsViaHeartBeat,
+                this::generateAndAssignSplits,
+                0,
+                pollTieringTableIntervalMs);
+    }
+
+    @Override
+    public void handleSplitRequest(int subtaskId, @Nullable String requesterHostname) {
+        if (!context.registeredReaders().containsKey(subtaskId)) {
+            // reader may be failed, skip this request.
+            return;
+        }
+        LOG.info("TiringSourceReader {} requests split.", subtaskId);
+        readersAwaitingSplit.add(subtaskId);
+        this.context.callAsync(
+                this::requestTieringTableSplitsViaHeartBeat, this::generateAndAssignSplits);
+    }
+
+    @Override
+    public void addSplitsBack(List<TieringSplit> splits, int subtaskId) {
+        readersAwaitingSplit.add(subtaskId);
+        pendingSplits.addAll(splits);
+        this.context.callAsync(
+                this::requestTieringTableSplitsViaHeartBeat, this::generateAndAssignSplits);
+    }
+
+    @Override
+    public void addReader(int subtaskId) {
+        LOG.info("Adding reader: {} to Tiering Source enumerator.", subtaskId);
+        if (context.registeredReaders().containsKey(subtaskId)) {
+            readersAwaitingSplit.add(subtaskId);
+        }
+    }
+
+    @Override
+    public void handleSourceEvent(int subtaskId, SourceEvent sourceEvent) {
+        // TODO: deal with finished table and failed table later via SourceEvent
+    }
+
+    private void generateAndAssignSplits(
+            Tuple3<Long, Long, TablePath> tieringTable, Throwable throwable) {
+        checkState(throwable == null, "Failed to request tiering table due to:", throwable);
```

Comment:
Maybe we can just warn this `throwable` instead of fail the whole job to make the job more stable? It's not fatal error, we can just try to request table in next round of tiering heart beats?